SalesforceCommerceCloud/pwa-kit
# Add webpack config plugin

(This is the new version for PR https://github.com/SalesforceCommerceCloud/pwa-kit/pull/234)

## Problem:

As we are adding feature to the SDK, we realize there is no intuitive way for developers to configure the SDK features, and there is no schema that helps developer understand them.

## Solution:

Created a config file `pwa-kit.config.json` at the root of the PWA project. This file contains all SDK configurations.

This PR adds a new webpack plugin `PwaKitConfigPlugin` to the SDK. The plugin is responsible for taking the configuration file and validate against the schema. The JSON schema can be imported from in SDK `pwa-kit-react-sdk/schemas/sdk-config.json`.

The webpack plugin taps into webpack `compilation` hook and will try to look for the file `pwa-kit.config.json`, if file not found, do nothing. If the file exists, validate it against the schema and show errors in the console.

We installed a new dependency in this change: `ajv`.

## Test:
- clone the repo if you haven't done so, run `npm ci` at root
- `cd packages/pwa` and then `npm start`
- the app should start as normal since the default config file has correct format.
- change `pwa-kit.config.json`, update `url` to have value of 1 (number instead of an object)
- verify you get console warning like this:
![Screen Shot 2021-12-09 at 2 23 46 PM](https://user-images.githubusercontent.com/10948652/145485285-1610aff5-8dc7-4e18-943a-df2a1a1e785e.png)


<img width="344" alt="Screen Shot 2021-12-09 at 2 23 42 PM" src="https://user-images.githubusercontent.com/10948652/145485292-95e6216c-52e1-49ae-a6f4-57463f37fb35.png">



- try to make the validation fail by changing the configuration values, e.g. remove required fields. you should see console errors.
- fix the configuration value, and the app server should restart and recover.
